src/widgets/popup-menu-button/PopupMenuButton.spec.tsx

```diff
@@ -0,0 +1,188 @@
+			<div key="buttonWrapper">
+				<button classes={css.button} key="popupMenuButton" onclick={noop}>
+					Button
+				</button>
+			</div>
+		</div>
+	));
+
+	const popupMenu = (
+		position: PopupMenuPositioning = {
+			right: 0,
+			top: 24
+		}
+	) => (
+		<PopupMenu
+			key="popupMenu"
+			position={position}
+			onFocus={noop}
+			onBlur={noop}
+			items={[
+				{
+					title: 'foo',
+					onClick: () => {}
+				}
+			]}
+		/>
+	);
+
+	beforeEach(() => {
+		jest.useFakeTimers();
+
+		const mockMetaMixin = new MockMetaMixin(PopupMenuButton);
+
+		mockMetaMixin.registerMetaCall(Dimensions, 'get', ['buttonWrapper'], {
+			size: {
+				width: 100,
+				height: 24
+			}
+		});
+
+		MockPopupMenuButton = mockMetaMixin.getClass();
+
+		itemClick = jest.fn();
+		items = [
+			{
+				title: 'foo',
+				onClick: itemClick
+			}
+		];
+	});
+
+	it('default renders correctly open', () => {
+		const h = harness(() => <MockPopupMenuButton items={[]}>Button</MockPopupMenuButton>);
+		h.expect(baseAssertion);
+	});
+
+	it('toggles on click', () => {
+		const h = harness(() => <MockPopupMenuButton items={items}>Button</MockPopupMenuButton>);
+
+		h.trigger('@popupMenuButton', 'onclick');
+
+		const assertion = baseAssertion.append(':root', () => [popupMenu()]);
+		h.expect(assertion);
+		h.trigger('@popupMenu', 'onFocus');
+
+		// Click should do nothing, but the removal of focus does
+		h.trigger('@popupMenuButton', 'onclick');
+
+		h.expect(assertion);
+
+		h.trigger('@popupMenu', 'onBlur');
+		h.expect(baseAssertion);
+
+		// Instally clicking again does not open the menu until after a 250ms timeout (after onBlur) to prevent the popup menu from flashing closed then opened
+		h.trigger('@popupMenuButton', 'onclick');
+		h.expect(baseAssertion);
+
+		jest.runAllTimers();
+
+		h.trigger('@popupMenuButton', 'onclick');
+		h.expect(assertion);
+	});
+
+	it('closes popup on blur', () => {
+		const h = harness(() => <MockPopupMenuButton items={items}>Button</MockPopupMenuButton>);
+
+		h.trigger('@popupMenuButton', 'onclick');
+		h.trigger('@popupMenu', 'onBlur');
+
+		h.expect(baseAssertion);
+	});
+
+	it('closes popup on menu item click', () => {
+		const h = harness(() => <MockPopupMenuButton items={items}>Button</MockPopupMenuButton>);
+
+		h.trigger('@popupMenuButton', 'onclick');
+		h.trigger('@popupMenu', (node) => {
+			return (node.properties as any).items[0].onClick;
+		});
+
+		expect(itemClick).toHaveBeenCalled();
+		h.expect(baseAssertion);
+	});
+
+	describe('positions', () => {
+		const testPosition = (positionName: PopupMenuPosition, position: PopupMenuPositioning) => {
+			const h = harness(() => (
+				<MockPopupMenuButton position={positionName} items={items}>
+					Button
+				</MockPopupMenuButton>
+			));
+
+			h.trigger('@popupMenuButton', 'onclick');
+
+			const assertion = baseAssertion.append(':root', () => [popupMenu(position)]);
+			h.expect(assertion);
+		};
+
+		test('top-left', () =>
+			testPosition('top-left', {
```

Comment:
Im really not the biggest fan of over abstracting tests, especially when both the harness and the assertion itself is moved  out. Given the actual abstraction is like 4 lines of code is it worth DRY for it? You also wouldn’t need to type the arguments for the function then I don’t think?